components/inputs/input-color.js

```diff
@@ -0,0 +1,411 @@
+					outline: none;
+					padding: 0.55rem;
+				}
+				.readonly-wrapper:focus {
+					outline: none;
+				}
+				.readonly-wrapper:${unsafeCSS(getFocusPseudoClass())} {
+					box-shadow: 0 0 0 2px #ffffff, 0 0 0 4px var(--d2l-color-celestine);
+				}
+			`
+		];
+	}
+
+	constructor() {
+		super();
+		this.disabled = false;
+		this.disallowNone = false;
+		this.labelHidden = false;
+		this.launchType = 'dialog';
+		this.readonly = false;
+		this.type = 'custom';
+		this._associatedValue = undefined;
+		this._missingLabelErrorHasBeenThrown = false;
+		this._opened = false;
+		this._validatingLabelTimeout = null;
+		this._value = undefined;
+	}
+
+	get associatedValue() { return this._associatedValue; }
+	set associatedValue(val) {
+		if (val !== undefined && !HEX_REGEX.test(val)) {
+			throw new TypeError(`<d2l-input-color>: invalid HEX associated-value "${val}"`);
+		}
+		const oldVal = this._associatedValue;
+		this._associatedValue = (typeof val === 'string') ? val.toUpperCase() : undefined;
+		this.requestUpdate('associatedValue', oldVal);
+	}
+
+	get value() {
+		if (this._value === undefined && this.disallowNone) {
+			return this.type === 'background' ? DEFAULT_VALUE_BG : DEFAULT_VALUE;
+		}
+		return this._value;
+	}
+	set value(val) {
+		if (val !== undefined && !HEX_REGEX.test(val)) {
+			throw new TypeError(`<d2l-input-color>: invalid HEX value "${val}"`);
+		}
+		const oldVal = this._value;
+		this._value = (typeof val === 'string') ? val.toUpperCase() : undefined;
+		this.requestUpdate('value', oldVal);
+	}
+
+	static get focusElementSelector() {
+		return '#opener';
+	}
+
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+		this._validateLabel();
+	}
+
+	render() {
+
+		const label = !this.labelHidden ? html`<div class="d2l-input-label">${this._getLabel()}</div>` : nothing;
+		const tooltip = !this._opened ? html`<d2l-tooltip for="opener" for-type="label">${this._getTooltipLabel()}</d2l-tooltip>` : nothing;
+		const opener = this._getOpener();
+
+		return html`${label}${opener}${tooltip}`;
+
+	}
+
+	async updated(changedProperties) {
+
+		super.updated(changedProperties);
+
+		if (changedProperties.has('label') || changedProperties.has('type')) this._validateLabel();
+		if (changedProperties.has('launchType')) this._validateLaunchType();
+
+	}
+
+	_getLabel() {
+		if (this.label !== undefined) return this.label;
+		if (this.type === 'background') return this.localize('components.input-color.backgroundColor');
+		if (this.type === 'foreground') return this.localize('components.input-color.foregroundColor');
+		return '';
+	}
+
+	_getOpener() {
+		if (this.readonly) {
+			return html`<div class="readonly-wrapper" id="opener" role="img" tabindex="0">${this._getSwatch()}</div>`;
+		}
+		const buttonClass = {
+			'd2l-button': true,
+			'd2l-dropdown-opener': this.launchType === 'dropdown',
+			'opened': this._opened
+		};
+		const ariaLabel = this._opened ? this._getTooltipLabel() : undefined;
```

Comment:
When the dropdown is open we hide the tooltip, so we need an explicit `aria-label`.